### PR TITLE
WIP: Initial implementation of gnu archives with symbols

### PR DESCRIPTION
**Pull request by GregBowyer**
# TL;DR
This is an attempt at building symbol tables, it is born out of using this crate inside an internal compiler which generates large and exotic archives complete with symbol tables. Aimed at solving #16 

It is a breaking change.

# How this works
Due to my requirements, this tries as best it can to use limited memory to write out the special archive sections for GNU archives. As such it uses temp files if an (end user configurable) in memory buffer would be too large. These files are spilled to disk and are assembled into the final archive, in the case of Unix (the platform our compiler works against) an attempt is made to copy the temp files into the final archive as efficiently as possible.

These are all hidden behind the `BaseBuilder` trait which implements `Builder` as a public trait for consumption.

Since each section is built as needed I removed the requirement to know all identifiers up front, they are added into the the name table as needed.

The symbol iterator is slightly extended to allow for symbol iteration. Mapping these back to entities is a little awkward right now as AFAIK the lack of GATS makes it impossible to express the lifetime of `Archive -> Iterator<Item=Symbol> -> Iterator<Item=Entry>`. Presently you have to consume the symbols into offsets and use those offsets to get the specific entries involved.

Its important to note that while a given `Symbol` represents a specific `Entry` its more than possible to have archives with the same symbol _name_ represented by multiple entries, that is that symbol name is many to many on archive entries. This shows up with weak symbols in compilation.

# Added tests
To test for validity I added the following tests:

* A basic integration test to make sure a GNU `ar` and `nm` tool understands the presented archive
* General tests around the building of archives
* Property tests for GNU archives.

# This collides with #17 
@bjorn3 I had this in progress for the last few weeks I am not 100% how to proceed as we both have taken a swing at the same problem. We should collaborate on what we want / need here I think.

# This is incomplete due to the following:

[ ] Needs implementation of builders for legacy and bsd archives
[ ] Needs reader implementation for `/SYM64` tables
[ ] Lacks a stress test for large archives
[ ] Probably should add basic benchmarks for this

### CODE REVIEW

File: src/builder/mod.rs [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+/// Archives are not expected to have gaps in them (and again some linkers object to this) and so
+/// its not possible to just make a file with a large pad between the special members and the rest
+/// of the resulting data.
+///
+/// Naturally the source code of LLVM and or binutils explains the format somewhat, but can be a
+/// bit obstrfucatory.
+///
+/// This mod encapsulates the generic parts of generating archives at scale but is aimed at hiding
+/// the implementation details from end users.
+pub(crate) mod private {
```
Comment (bjorn3) [inline src/builder/mod.rs:174]: No need to wrap it in `mod private` if you never use `pub`.

File: src/lib.rs [hunk 2]
```diff
@@ -63,26 +65,34 @@
...
 use std::os::unix::fs::MetadataExt;
 
-#[cfg(unix)]
-use std::os::unix::ffi::OsStrExt;
+mod error;
+pub mod builder;
+pub use builder::{Builder, GnuBuilder};
+
+/// Internal optimisation of idents, assuming most idents are smaller than the extended
+/// string size of the formats that allow such strings.
```
Comment (bjorn3) [inline src/lib.rs:93]: For rust this is not the case. Most identifiers are like twice the size. Eg `rust_out.rust_out.7rcbfp3g-cgu.0.rcgu.o` is 40 chars.

File: src/lib.rs [hunk 3]
```diff
@@ -136,23 +164,18 @@ pub struct Header {
...
-    /// other fields set to zero.
-    pub fn new(identifier: Vec<u8>, size: u64) -> Header {
+    pub(crate) fn new_with_id(identifier: Ident, size: u64) -> Self {
         Header {
-            identifier,
             mtime: 0,
             uid: 0,
             gid: 0,
-            mode: 0,
+            mode:0o644,
```
Comment (bjorn3) [inline src/lib.rs:172]: nit:

```suggestion
            mode: 0o644,
```

File: src/lib.rs [hunk 4]
```diff
@@ -693,6 +694,27 @@ impl<R: Read + Seek> Archive<R> {
         })
     }
 
+    /// Get the entry indexes for the given symbol name
+    ///
+    /// These can be used later with `jump_to_entry` to get the actual entry
+    pub fn entries_for_symbol<S: AsRef<[u8]>>(&mut self, symbol: S) -> io::Result<SymbolEntries> {
```
Comment (bjorn3) [inline src/lib.rs:700]: This makes it very easy for accidental quadratic runtime to happen if you call this in a loop for every symbol you want.

File: src/lib.rs [hunk 5]
```diff
@@ -693,6 +694,27 @@ impl<R: Read + Seek> Archive<R> {
...
+        self.scan_if_necessary()?;
```
Comment (bjorn3) [inline src/lib.rs:701]: Doesn't `self.symbols()` already call this?

File: tests/fixtures/object_normal_syms.c [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#include "object_syms.h"
```
Comment (bjorn3) [inline tests/fixtures/object_normal_syms.c:1]: Please include the exact compiler invocation and version to reproduce these files. Including which Linux distro. (They often patched compilers slightly, which may affect the output)

File: tests/gnu_ar.rs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    let gnu_out = gnu_ar.arg("tv").arg(test_file.path().as_os_str()).output()?;
+    assert!(gnu_out.status.success());
+    assert_eq!(indoc!(r#"
+        rw-r--r-- 0/0   1328 Jan  1 00:00 1970 object_normal_syms.o
+        rw-r--r-- 0/0   1232 Jan  1 00:00 1970 object_weak_syms.o
+    "#), as_str(&gnu_out.stdout));
+
+    // NM test here
+    if let Some(mut gnu_nm) = command("nm") {
+        let nm_out = gnu_nm.arg("--print-armap").arg(test_file.path().as_os_str()).output()?;
```
Comment (bjorn3) [inline tests/gnu_ar.rs:67]: While creating my PR, I discovered that for archives without symbol table `ar` just reads the object files.

File: examples/symbols.rs [hunk 8]
```diff
@@ -26,6 +26,6 @@ fn main() {
     let mut archive = ar::Archive::new(input_file);
 
     for symbol in archive.symbols().expect("failed to parse symbols") {
-        println!("{}", String::from_utf8_lossy(symbol));
+        println!("{}", symbol.name().expect("symbol name is not UTF-8"));
```
Comment (bjorn3) [inline examples/symbols.rs:29]:
```suggestion
        println!("{}", String::from_utf8_lossy(symbol.raw_name());
```

File: src/builder/gnu.rs [hunk 9]
```diff
@@ -0,0 +1,716 @@
...
+//!
+//! 2) The handling of extended strings is done with a string lookup table (either as the first of
+//!    second member) which is little more than a large string array.
+//!
+//! 3) Extensions exist to create a rare format known as a thin-archive.
+//!
+//! 4) GNU archives have a formal [deterministic mode](#deterministic-archives) that is important
+//!    for build systems and toolchains.
+//!
+//! Most tools outside of BSD targets tend to use GNU format as the defacto standard, and it is
```
[the remark this answers is not in the document]
Comment (GregBowyer) [reply src/builder/gnu.rs:22]: Solaris and if you stretch the definition windows too

File: src/lib.rs (hunk 4 above)
Comment (GregBowyer) [reply src/lib.rs:700]: Yes :(

I hate this part of the API and would love suggestions for a better API.
Since its a landmine I guess we could take it away and just have people deal with it themselves ?

File: tests/fixtures/object_normal_syms.c (hunk 6 above)
Comment (GregBowyer) [reply tests/fixtures/object_normal_syms.c:1]: I dont think it will matter _too_ much, its more to have something with real symbols for `nm --print-armap` since it checks what we claim as a symbol vs the entry in the archive.

We could generate these as part of `build.rs`, or they could be rust objects

File: tests/gnu_ar.rs (hunk 7 above)
Comment (GregBowyer) [reply tests/gnu_ar.rs:67]: Yep thats why the `nm` check. It caught me out more than once too.

Should we add a comment here?

File: src/lib.rs (hunk 2 above)
Comment (GregBowyer) [reply src/lib.rs:93]: Humm if that's the case then we could go back to Vec I think

Having an online smallvec of 40bytes seems to be too large to take advantage of.

Wdyt? I say we kill Ident since for your usecases it's always going to be too large to take advantage of an online vec

File: src/builder/gnu.rs [hunk 10]
```diff
@@ -0,0 +1,716 @@
...
+
+    fn compute_metadata_params(&mut self) -> Result<MetadataParams> {
+        let num_syms = self.symbols.len() as u64;
+
+        let mut metadata_size = GLOBAL_HEADER_LEN as u64;
+
+        // Calculate symtab size if we are adding one
+        if !self.symbols.is_empty() {
+            metadata_size = metadata_size
+                .saturating_add(ENTRY_HEADER_LEN as u64)
```
[the remark this answers is not in the document]
Comment (GregBowyer) [reply src/builder/gnu.rs:210]: Overflow in the 64bit case should be an error, but overflow at 32bits is needed to let us know when to produce a `/SYM64` table.

Will rework to only care about the 64bit overflow

File: src/builder/mod.rs [hunk 11]
```diff
@@ -0,0 +1,446 @@
...
+    let identifier = path.as_ref()
+        .file_name()
+        .map(|name| name.encode_wide())
+        .fold(Ident::new(), |(mut ident, wide)| {
+            // Little-endian
+            ident.push((wide & 0xFF) as u8);
+            ident.push((wide >> 8) as u8);
+            ident
+        })
+        .ok_or_else(|| err!("Given path doesn't have a file name"))?;
```
[the remark this answers is not in the document]
Comment (GregBowyer) [reply src/builder/mod.rs:51]: This was the original code I started with, I dont profess to know enuff about the windows toolchain to honestly answer this.

Is there a reference for windows?

**Comment by bjorn3** [comment]
@GregBowyer and I have talked in DM about a plan forward.
